File: aoc/aoc2019/day03/Day3.fs
module Day3
open System.IO
open System
open System.Collections.Generic

type Day3() =
    static member Run() =
        use file = new StreamReader("day03/p.in")

        let rec parseLine (lineNo: int) (l1: (char * int) list) (l2: (char * int) list) =
            match file.ReadLine() with
            | null -> (l1, l2)
            | ln ->
                let parts = ln.Split(',') |> Array.toList
                let dirLengthTuple (part: string) : char * int =
                    (part.[0], int32 (part.Substring(1)))
                let l1', l2' =
                    if lineNo = 0 then
                        List.fold (fun acc part -> dirLengthTuple part :: acc) l1 parts |> List.rev, l2
                    else
                        l1, List.fold (fun acc part -> dirLengthTuple part :: acc) l2 parts |> List.rev
                parseLine (lineNo + 1) l1' l2'

        let l1, l2 = parseLine 0 [] []

        let mutable closest = Int32.MaxValue
        let visited = HashSet<Tuple<int, int>>()
        let dic = Dictionary<Tuple<int, int>, int>()

        let mutable current = (0, 0)
        let mutable steps = 1
        for i = 0 to List.length l1 - 1 do
            let dir, len = List.item i l1
            for _ = 0 to len - 1 do
                match dir with
                | 'U' -> current <- (fst current, snd current - 1)
                | 'D' -> current <- (fst current, snd current + 1)
                | 'L' -> current <- (fst current - 1, snd current)
                | 'R' -> current <- (fst current + 1, snd current)
                visited.Add(current)
                if not (dic.ContainsKey current) then
                    dic.Add(current, steps)
                steps <- steps + 1

        current <- (0, 0)
        steps <- 1
        for i = 0 to List.length l2 - 1 do
            let dir, len = List.item i l2
            for _ = 0 to len - 1 do
                match dir with
                | 'U' -> current <- (fst current, snd current - 1)
                | 'D' -> current <- (fst current, snd current + 1)
                | 'L' -> current <- (fst current - 1, snd current)
                | 'R' -> current <- (fst current + 1, snd current)
                if visited.Contains current then
                    if steps + dic.[current] < closest then
                        closest <- steps + dic.[current]
                steps <- steps + 1

        Console.WriteLine(closest)
        file.Close()

File: aoc/aoc2022/day03/Day3.cs
public class Day3
{
    public static void Run()
    {
        using (StreamReader file = new StreamReader("day03/p.in"))
        {
            var prios = 0;
            var dic = new Dictionary<char, int>();
            var index = 0;

            string ln;
            while ((ln = file.ReadLine()) != null)
            {
                for (var i = 0; i < ln.Length; i++)
                {
                    if (dic.ContainsKey(ln[i]))
                    {
                        if (dic[ln[i]] == index - 1)
                        {
                            if (index == 2)
                            {
                                if (ln[i] >= 'a' && ln[i] <= 'z') prios += ln[i] - 'a' + 1;
                                if (ln[i] >= 'A' && ln[i] <= 'Z') prios += ln[i] - 'A' + 27;
                                dic.Clear();
                                break;
                            }
                            else
                            {
                                dic[ln[i]] = index;
                            }
                        }
                    }
                    else if (index == 0)
                    {
                        dic[ln[i]] = index;
                    }
                }
                index = (index + 1) % 3;
            }
            file.Close();

            Console.WriteLine(prios);
        }
    }
}

